packages/sql/src/RequestManager.ts

```diff
@@ -0,0 +1,167 @@
+import { Credentials } from './credentials';
+
+type PromiseCb<T> = (value?: T) => void;
+type FetchArgs = [PromiseCb<any>, PromiseCb<any>, RequestInfo, RequestInit?];
+
+const UNKNOWN = -1;
+
+export class RequestManager {
+  private _username: string;
+  private _apiKey: string;
+  private _server: string;
+  private _queue: FetchArgs[];
+  private _callsLeft: number = UNKNOWN;
+  private _retryAfter: number = UNKNOWN;
+  private _retryTimeoutId: number = UNKNOWN;
+  private _fetching: boolean = false;
+  private _scheduleDebounce: number = UNKNOWN;
+
+  constructor(credentials: Credentials) {
+    const { username, apiKey, server } = credentials;
+
+    this._username = username;
+    this._apiKey = apiKey;
+    this._server = server.replace('{user}', username);
+    this._queue = [];
+  }
+
+  protected set callsLeft(value: number) {
+    this._callsLeft = value;
+  }
+
+  protected get callsLeft() {
+    return this._callsLeft;
+  }
+
+  protected _scheduleRequest(
+    resolve: PromiseCb<any>,
+    reject: PromiseCb<any>,
+    info: RequestInfo,
+    init?: RequestInit) {
+
+    this._queue.push([resolve, reject, info, init]);
+
+    clearTimeout(this._scheduleDebounce);
+    this._scheduleDebounce = window.setTimeout(() => {
+      this._scheduler();
+    }, 0);
+  }
+
+  private _scheduler() {
+    if (this._queue.length === 0) {
+      return;
+    }
+
+    if (this._retryTimeoutId !== UNKNOWN || this._fetching) {
+      return;
+    }
+
+    if (this._retryAfter !== UNKNOWN) {
+      this._retryTimeoutId = window.setTimeout(() => {
+        this._retryTimeoutId = UNKNOWN;
+        this._retryAfter = UNKNOWN;
+        // Technically? But this is risky.
+        this._callsLeft += 1;
+        this._scheduler();
+      }, this._retryAfter * 1000);
+
+      return;
+    }
+
+    const nRequests = this._callsLeft !== -1 ?
```

Comment:
I would suggest a bunch of smaller methods here. It is not urgent now, but I'm gonna write them in any case, for a future refactor (after merging):

* From 60 to 66: a new small method like `private _setRetry()`
* From 71 to 84--> `private getRequestPromises()`
* From 86 to 93 --> another method to deal with previous promises and the queue (not sure about the naming)